grblas/base.py

```diff
@@ -125,13 +233,128 @@ def new(self, *, dtype=None, mask=None):
+        if self.resolved_indexes.is_single_element():
+            if mask is not None:
+                raise TypeError('mask is not allowed for single element extraction')
+            val, cur_dtype = self.parent._extract_element(self.resolved_indexes)
+            if dtype is None:
+                dtype = cur_dtype
+            from .scalar import Scalar
+            return Scalar.new_from_value(val, dtype=dtype)
+        else:
+            delayed_extractor = self.parent._prep_for_extract(self.resolved_indexes)
+            return delayed_extractor.new(dtype=dtype, mask=mask)
+
+    def _extract_delayed(self):
+        """Return a GbDelayed object, treating this as an extract call"""
+        if isinstance(self.parent, Updater):
+            raise TypeError('Cannot extract from an Updater')
+        if self.resolved_indexes.is_single_element():
+            raise TypeError('extract and update is not allowed for single element')
+        return self.parent._prep_for_extract(self.resolved_indexes)
+
+
+class IndexerResolver:
+    def __init__(self, obj, indices):
+        if obj.is_scalar:
+            raise TypeError("Cannot index into Scalars")
+        self.obj = obj
+        self.indices = self.parse_indices(indices, obj.shape)
+
+    def is_single_element(self):
+        for idx, size in self.indices:
+            if size is not None:
+                return False
+        return True
+
+    def parse_indices(self, indices, shape):
+        """
+        Returns
+            [(rows, rowsize), (cols, colsize)] for Matrix
+            [(idx, idx_size)] for Vector
+
+        Within each tuple, if the index is of type int, the size will be None
+        """
+        if len(shape) == 1:
+            if type(indices) == tuple:
+                raise TypeError(f'Index for {self.obj.__class__.__name__} cannot be a tuple')
+            # Convert to tuple for consistent processing
+            indices = (indices,)
+        elif len(shape) == 2:
+            if type(indices) != tuple or len(indices) != 2:
+                raise TypeError(f'Index for {self.obj.__class__.__name__} must be a 2-tuple')
+
+        out = []
+        for i, idx in enumerate(indices):
+            typ = type(idx)
+            if type == tuple:
+                raise TypeError(f'Index in position {i} cannot be a tuple; must use slice or list or int')
+            out.append(self.parse_index(idx, typ, shape[i]))
+        return out
+
+    def parse_index(self, index, typ, size):
+        if typ == int:
+            if index >= size:
+                raise IndexError(f'index={index}, size={size}')
+            return index, None
+        if typ == slice:
+            if index == slice(None):
+                # [:] means all indices; use special GrB_ALL indicator
+                return lib.GrB_ALL, size
+            index = tuple(range(size)[index])
+        elif typ != list:
+            try:
+                index = tuple(index)
```

Comment:
What is `index` likely to be here, and what errors might we expect?  Do we need to inspect the elements?

Reply:
Elements should be:

- int for a single index
- slice for many indices
- list for many indices

I'm willing to treat other iterables (except tuples) as valid input. If that's too lenient, I'm willing to be more strict for the sake of safety and predictable behavior.

Reply:
`pandas` and `numpy` are pretty strict.  We can always start strict and relax it later if we find the flexibility is worth it.